src/controllers/UserController.ts

```diff
@@ -0,0 +1,115 @@
+import {
+  Body,
+  Controller,
+  Path,
+  Post,
+  Response,
+  Route,
+  SuccessResponse,
+  Tags,
+} from "tsoa";
+import type {
+  AddOrUpdateUserRequest,
+  AddOrUpdateUserResponse,
+  ApiResponse,
+} from "../types/api.js";
+import { z } from "zod";
+import { getEvmClient } from "../utils/getRpcUrl.js";
+import { SupabaseDataService } from "../services/SupabaseDataService.js";
+
+@Route("v1/users")
+@Tags("Users")
+export class UserController extends Controller {
+  /**
+   * Add or update a user
+   */
+  @Post(`{address}`)
+  @SuccessResponse(201, "User updated successfully", "application/json")
+  @Response<ApiResponse>(422, "Unprocessable content", {
+    success: false,
+    message: "Errors while validating user",
+  })
+  public async addOrUpdateUser(
+    @Path() address: string,
+    @Body() requestBody: AddOrUpdateUserRequest,
+  ): Promise<AddOrUpdateUserResponse> {
+    const inputSchema = z.object({
+      display_name: z.string().optional(),
+      avatar: z.string().optional(),
+      signature: z.string(),
+      message: z.string(),
+    });
+    const parsedBody = inputSchema.safeParse(requestBody);
+    if (!parsedBody.success) {
+      this.setStatus(400);
+      return {
+        success: false,
+        message: "Invalid input",
+        data: null,
+        errors: JSON.parse(parsedBody.error.toString()),
+      };
+    }
+
+    const { signature, message } = parsedBody.data;
```

Comment:
What is in the signed message? Does it contain a chainId? Based on line 54 it looks like this is fixed to Optimism, but we should be able to handle different chains without introducing breaking changes. This entail that the message should contain a chainId to a) prevent replays and b) support account abstraction methods like Safe

Reply:
To be clear, this never hits the chain so I'm not sure if replays are a thing (if I understand correctly)? Only if the request would be intercepted, you could change somebody profile info again but that would still be possible when adding a chain id to the message body 🤔 

How do you envision this. AFAICT there's 3 options:

- Sign a message like `${chainId} - ${address} - Update user settings`.
- Sign typed data such as `{ chainId: number; address: string; message: "Update user settings"
- Simply send chainId in the request as well and initialize the client with that.

Reply:
[Went through this comment and the links posted](https://discord.com/channels/1075404472152494100/1280134960212217898/1288039895750676553) and seems like typed data would be the option.

Reply:
Implemented this in https://github.com/hypercerts-org/hypercerts-api/pull/168/commits/29b5e9c348349660a3d2bf0dcb395672df1b64fa, @bitbeckers please let me know if you agree with the implementation.

Reply:
Thanks. As it follows EIP712, it need the DomainSeparator as defined here: https://eips.ethereum.org/EIPS/eip-712#definition-of-domainseparator

In our case, only field `name, version, chainId` are applicable. This lead to having two fields names `chainId` and `address`, as the `signer` is the `address` and `chainId` is in the DomainSeparator.

Is there a case where we want address and chainId explicitly declared outside of the DomainSeparator?